### PR TITLE
Use napari-console package

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR begins to address #1593 by making us use [napari-console](https://github.com/sofroniewn/napari-console) instead of creating the console internally. That package isn't released on pypi yet, so for this to work you'll need to clone it. We can work on cleaning it up and getting the first release out.

Right now this PR uses napari-console as an external package and doesn't use our hookspec. One nice things is that lets us keep the console imports lazy, keeps the special short cut etc.

I think it probably makes sense to get napari-console `0.0.1` released without the hookimpls exposed and using this dependency approach as it is 100% backwards compatible. We could then move to a hookimpl approach and/or making the install optional. Curious what @jni @tlambert03 recommend? Contributions to napari-console directly are welcome too. Once we have the `0.0.1` (or now if people think it is close enough) we can move over to the napari org instead of my personal account. We might want to clean up some stuff like tests/ readme before doing that though.

## Type of change

- [x] Refactor (non-breaking change which fixes an issue)

### CODE REVIEW

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [inline napari/_qt/qt_viewer.py:300]: i assume the update theme stuff just isn't necessary anymore to keep styles in sync?

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [inline napari/_qt/qt_viewer.py:300]: maybe a warning with instructions on how to install the console?

**Review by tlambert03 — COMMENTED** [review]
This looks fine to me @sofroniewn.  Nice to remove a few direct dependencies! :)  Is the code basically just moved to napari-console?  or is it edited there yet?

I'd say let's just go ahead and put it in napari/ org ...

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:300]: ah so it's interesting, in the new package i connect to the viewer theme event internally, there are some pros/ cons around that so we might to think about if there is a cleaner way

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:300]: very nice idea

**Comment by sofroniewn** [comment]
> Is the code basically just moved to napari-console? or is it edited there yet?

It's basically just moved, a couple minor changes around theme updates, accepting the viewer as an argument, but not much different

> I'd say let's just go ahead and put it in napari/ org ...

i'll do that right now. we can then keep convos going there too.

**Comment by sofroniewn** [comment]
Repo now transferred https://github.com/napari/napari-console, and everyone in the napari org should now have commit rights etc.

**Comment by jni** [comment]
One thing I'll point out is that simul-releasing lots of packages on conda-forge is annoying. So ideally I'd like to pull this bandage all at once and make the console optional in a single release. Notifying users of this might be tricky though... 😕 Possible idea:

(1) release a shim `napari-console` package that doesn't do anything. In 0.4.4, turning on the console or accessing the attribute *without* napari-console installed will raise a warning inside the console that they'll need to install napari-console as well to keep using it starting 0.4.5. If they install it, the message is silenced.
(2) between 0.4.4 and 0.4.5 we remove the console by default and put all that code in napari-console

Thoughts?

**Comment by sofroniewn** [comment]
> (2) between 0.4.4 and 0.4.5 we remove the console by default and put all that code in napari-console

This has already been done.

> (1) release a shim napari-console package that doesn't do anything.

Given that the code for the console is already in the repo we might as well just release it. I was actually thinking I might even have that done over the weekend but I didn't quite get to it. We can see. One idea is to make the initial release without the hookspec.

> One thing I'll point out is that simul-releasing lots of packages on conda-forge is annoying. So ideally I'd like to pull this bandage all at once and make the console optional in a single release. Notifying users of this might be tricky though... 😕

It's a good point, important to note, but I think we can get a staggered release fairly easily, first getting napari-console out, then getting stuff into napari. We shouldn't actually need to make breaking change updates to napari-console after that so I'm not so concerned about simul-releasing.

Let's see how it goes - I'm all about looking for the lowest friction path here, and if I'm missing something then we can go with the shim package and warnings approach!

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:300]: I've added a warning, let me know if it was what you had in mind/ if you think text can be improved

**Comment by sofroniewn** [comment]
Ok, we've now released the `0.0.1` version of napari-console on pypi. I know @jni mentioned conda building simul-release difficulties. Do we need to make a conda-forge package for napari-console now for `0.0.1`?

`napari-console` right now is not using the add_dock_widget hookspec. I thought it would be easier to pull out the functionality first, nor is it optional yet.

Curious @tlambert03 how you would recommend we proceed at this point?

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [inline napari/_qt/qt_viewer.py]:
```suggestion
                    ' "pip install napari_console"'
```
(note additional space there too)

**Review by tlambert03 — APPROVED** [review]
just tried it out, works great!

> napari-console right now is not using the add_dock_widget hookspec. I thought it would be easier to pull out the functionality first, nor is it optional yet.

that's fine with me

**Comment by tlambert03** [comment]
> Do we need to make a conda-forge package for napari-console now for 0.0.1?

I think we should get on it right away.  I think it'd be nice to make a new napari release in the next week or so with some of the updates we're discussing (particularly if we decide to modify the new hookspec APIs at all), and this will block us if we don't have it

**Comment by sofroniewn** [comment]
> I think we should get on it right away. I think it'd be nice to make a new napari release in the next week or so with some of the updates we're discussing (particularly if we decide to modify the new hookspec APIs at all), and this will block us if we don't have it

ok so sounds like we need to just get napari-console released on conda-forge. I don't know how to do that. @jni can you take lead on that?

Also a comment that right now napari-console isn't really usable without napari, but it doesn't depend on either napari or Qt. I think that's fine. There is no need for it to be "standalone" but i just wanted to comment on that

**Comment by sofroniewn** [comment]
The conda-forge recipe is here https://github.com/conda-forge/staged-recipes/pull/13773

Once this PR is merged I'll update this PR to use the 0.0.2 version of napari-console and then we can merge

**Comment by sofroniewn** [comment]
Ok, the conda forge `napari-console` is now released and this setup cfg depends on `>0.0.2`. I'll wait to see if @jni or @tlambert03 have any final comments and if not will then merge promptly.

The next step is probably a follow-up PR to make this an optional dependency.

Then the final step is to switch the discovery mechanism over to using our plugin interface and a hookimpl (which will require the 0.0.3 napari-console release)

**Review by tlambert03 — APPROVED** [review]
looks good! tried it out.  works for me

**Pull request merged**
